### ISSUE TITLE
Settings: site name and tagline are sometimes reset to default values.

### ISSUE THREAD

**Issue by worldomonation** [Automattic/wp-calypso#64271]
### Quick summary

This has been observed occasionally in the context of end-to-end tests in CI, where the site tagline and site names are reset to the default values despite custom values having been entered during the onboarding flow.

Example site: 207546799.

### Steps to reproduce

1. the `/start` flow is visited.
2. user is created.
3. site is created with a custom `site_name` and `site_tagline` attribute.
4. user enters dashboard.
5. user checks Settings > General Settings.

### What you expected to happen

The custom site name and site tagline set in Step 3 are retained.

### What actually happened

The custom values are wiped out, replaced with default values.

![image](https://user-images.githubusercontent.com/6549265/171513967-ee17ff4a-31cd-4482-8963-a2bd9bf9b21c.png)

https://user-images.githubusercontent.com/6549265/171513975-f1c57a55-4866-4dd3-9a99-42d936a11cdb.mp4

### Context

CI automated testing.

p1654118344116219-slack-C02DQP0FP

### Other notes

Looking at the trace file generated by Playwright it has captured the following:

1. blog name field is filled with content eg. `e2eflowtesting1989712233`
2. once the focus shifts to the `Continue` button a new request is automatically sent:
```
{
  "code": 200,
  "headers": [
    {
      "name": "Content-Type",
      "value": "application/json"
    }
  ],
  "body": {
    "updated": []
  }
}
```
which generates the response
```
{"blogname":"Site Title","blogdescription":""}
```

### Reproducibility

Intermittent

**Comment by worldomonation** [comment]
It looks like these are the truthes of this screen:

![image](https://user-images.githubusercontent.com/6549265/172670111-89388524-411e-4731-bbb7-1b0912eb523b.png)

- from Intent Capture, clicking `Write` to enter the screen in question does not trigger a network request.
- the input fields are momentarily disabled when the page is loaded.
- waiting for `networkidle` event does not provide sufficient buffer.

**Comment by worldomonation** [comment]
After discussion with @dpasque we know a bit more about the behavior of the Writing screen.

When the site metadata fields load, they _appear_ to be disabled, but in reality they are not. Therefore, when the page initially loads, Playwright locates input fields, checks whether it is disabled (to which the response is they are not) and proceeds to set the fields with test data.

However, once the loading process is finished the defaults overwrite all the test data that was set.

The reason why Site Tagline often survives is likely because the first input to be filled is the Blog Name and the brief delay before moving on to Site Tagline is enough of a delay for the full load to complete.

@dpasque suggested waiting not only for the selector for the inputs, but also the default values, as they only come in after the initial load is finished.

**Comment by worldomonation** [comment]
Immediately after https://github.com/Automattic/wp-calypso/pull/64458 was merged into `trunk`, we see the spec fail again:

![image](https://user-images.githubusercontent.com/6549265/172732189-55ce6f4a-29b8-414a-9327-611ba1b2391c.png)

**Comment by dpasque** [comment]
Ah-ha! I've finally figured this one out!

It looks like this was likely caused by #64246 , specifically, adding [this effect hook](https://github.com/Automattic/wp-calypso/commit/190ab1b80d9a25a50e445aed992a900fa2aae160#diff-089e58ab4a377e7559811bda4c20e1d9ea848562b80acee5fb59e6a1fa25a93bR32-R40).

```typescript
	useEffect( () => {
		if ( ! site ) {
			return;
		}

		setSiteTitle( site.name || '' );
		setTagline( site.description );
	}, [ site ] );
```

That also lines up when we started seeing this pre-release test start to act flakey

(Pinging @markbiek so he can weigh in.) 👋 

What's happening in this race condition...
1. An API request is fired off to `/rest/v1.1/sites/<site_id>`
2. The UI of the component is rendered.
3. If the network is slow (e.g. 3G, or any other reason), the component UI will be interactive before the web request returns.
4. The user, or E2E tests, can update the field for site name, setting the local state of `siteTitle`
5. The web request finally comes back, updating the site details in the Redux store, triggering a re-render, and triggering this effect hook. This effect hook is then overwriting the local state with the value from the redux store/web request. This causes the "reset" looking behavior.

For a fix, it might be naive, but I _think_ we can just add a check in that effect hook to make sure we only set the site title/tagline if they are empty. That means we aren't overwriting any new values. That could still leave a condition though where someone clears out a value, then sees it pop back in, but that feels less disruptive. Alternatively, we could create a new piece of local state, a "dirty flag" boolean of sorts, that we set if the user has used the input elements at all. Then we would check for that in the effect hook, and not update from the redux store if that value is true.

Otherwise, we could make it so the component isn't interactive until the web request returns, but that feels kinda yucky.

Thoughts @markbiek ? This is an area I haven't really done any work in, so I want to make sure that fix wouldn't cause anything too disruptive!

**Comment by markbiek** [comment]
@worldomonation Thanks for the screenshare to show me the issue.

And extra thanks to you and @dpasque for all the details!

I agree that checking the title & tagline in the hook seems like the best way to fix it. I'll give that a shot and see how it goes.
